### PR TITLE
Added Practitioner class

### PR DESCRIPTION
**Pull request by that-guy-is-it**

### CODE REVIEW

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import lombok.NoArgsConstructor;
+
+@Data
+@Builder
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+@AllArgsConstructor
+@JsonAutoDetect(
+    fieldVisibility = JsonAutoDetect.Visibility.ANY,
+    isGetterVisibility = JsonAutoDetect.Visibility.NONE)
+@Schema(description = "https://www.hl7.org/fhir/practitioner.html")
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:34]: US Core R4 practitioner: http://hl7.org/fhir/us/core/StructureDefinition-us-core-practitioner.html

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  @Pattern(regexp = Fhir.ID)
+  String id;
+
+  @Valid Meta meta;
+
+  @Pattern(regexp = Fhir.URI)
+  String implicitRules;
+
+  @Pattern(regexp = Fhir.CODE)
+  String language;
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:46]: missing `text`, `contained`, `extension`, `modifierExtension`

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+  @Valid List<Identifier> identifier;
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:48]: not empty

Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:48]: `system` and `value` are required fields inside this `Identifier`, although vanilla R4 does not require them. @ilaflamme-va I think you ran into a similar problem recently? Should this class have its own `Identifier` class with more strict validation?

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+  Boolean active;
+
+  @Valid List<HumanName> name;
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:52]: not empty

Comment (jblefkove-va) [reply us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:52]: similar to above, `family` is a required field for `HumanName` on this class, but not for general R4

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  String date;
+
+  @Valid List<Attachment> photo;
+
+  @Valid List<Qualification> qualification;
+
+  @Valid List<CodeableConcept> communication;
+
+  enum GenderCode {
+    MALE,
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:70]: lowercase

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    @Pattern(regexp = Fhir.ID)
+    String id;
+
+    @Valid List<Extension> extension;
+
+    @Valid List<Extension> modifierExtension;
+
+    @Valid List<Identifier> identifier;
+
+    @NotNull CodeableConcept code;
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:93]: `@Valid`

**Comment by jblefkove-va** [comment]
Missing `Bundle` and `Entry` classes

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java (hunk 3 above)
Comment (jblefkove-va) [reply us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:48]: Found it: https://github.com/department-of-veterans-affairs/health-apis-fhir-resources/pull/99/files

So this class can use the shared R4 classes for `Identifier` and `HumanName`. The additional validations can be specified in custom validation methods on this class.

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+  @Valid List<Attachment> photo;
+
+  @Valid List<Qualification> qualification;
+
+  @Valid List<CodeableConcept> communication;
+
+  @JsonIgnore
+  @SuppressWarnings("unused")
+  @AssertTrue(message = "At most one IdentifierClia can be specified.")
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:90]: this message is not accurate

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+  private boolean isValidIdentifier() {
+    if (identifier == null) {
+      return false;
+    }
+    return identifier.stream().filter(e -> e.system() != null && e.value() != null).count() >= 1;
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:95]: should verify that *every* identifier has system and value set

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  @JsonIgnore
+  @SuppressWarnings("unused")
+  @AssertTrue(message = "At most one IdentifierClia can be specified.")
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:100]: inaccurate message

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+  private boolean isValidName() {
+    if (name == null) {
+      return false;
+    }
+    return name.stream().filter(e -> e.family() != null).count() >= 1;
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:105]: similarly this should verify that every name has `family `set

**Comment by jblefkove-va** [comment]
Needs a unit test, plus tests for the custom validations

File: us-core-r4/src/test/java/gov/va/api/health/r4/api/samples/SamplePractitioners.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+import gov.va.api.health.r4.api.resources.Practitioner;
+import java.util.Collections;
+import java.util.List;
+import lombok.NoArgsConstructor;
+import lombok.experimental.Delegate;
+
+@NoArgsConstructor(staticName = "get")
+public class SamplePractitioners {
+
+  @Delegate SampleDataTypes dataTypes = SampleDataTypes.get();
```
Comment (jblefkove-va) [inline us-core-r4/src/test/java/gov/va/api/health/r4/api/samples/SamplePractitioners.java:12]: Remove this if it is not used

File: us-core-r4/src/test/java/gov/va/api/health/r4/api/resources/PractitionerTest.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+  @Test
+  public void validationPassesGivenGoodIdentifier() {
+    Practitioner prac = data.practitioner();
+    assertThat(violationsOf(prac)).isEmpty();
+    prac.identifier().stream()
+        .forEach(
+            i -> {
+              i.value("value");
+              i.system("system");
```
Comment (jblefkove-va) [inline us-core-r4/src/test/java/gov/va/api/health/r4/api/resources/PractitionerTest.java:85]: you can chain these together... `i.value(...).system(...)`

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+
+  @Valid Narrative text;
+
+  @Valid List<SimpleResource> contained;
+
+  @Valid List<Extension> extensions;
+
+  @Valid List<Extension> modifierExtensions;
+
+  @Valid @NotEmpty List<Identifier> identifier;
```
Comment (ilaflamme-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:71]: missing validation method for an Identifier:NPI slice.

**Review by ilaflamme-va — CHANGES_REQUESTED** [review]

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java (hunk 13 above)
Comment (ilaflamme-va) [reply us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:71]: Your current validation method will throw away good Identifier:NPI slices

File: us-core-r4/src/test/java/gov/va/api/health/r4/api/samples/SamplePractitioners.java (hunk 11 above)
Comment (that-guy-is-it) [reply us-core-r4/src/test/java/gov/va/api/health/r4/api/samples/SamplePractitioners.java:12]: It is used, though. The delegate is how I get access to the constructors of the types, such as `meta()` and `narrative()`. Is there another way?

Comment (jblefkove-va) [reply us-core-r4/src/test/java/gov/va/api/health/r4/api/samples/SamplePractitioners.java:12]: My mistake, I didn't see references to it but missed the `@Delegate`

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+  @AssertTrue(message = "At most one IdentifierNpi can be specified.")
+  private boolean isValidIdentifierNpiSlice() {
+    if (identifier == null) {
+      return true;
+    }
+    return identifier.stream()
+            .filter(
+                e ->
+                    StringUtils.isNotBlank(e.system())
+                        && e.system().equals("http://hl7.org/fhir/sid/us-npi"))
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:103]: how about `"http://hl7.org/fhir/sid/us-npi".equals(e.system())` ?

Comment (that-guy-is-it) [reply us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:103]: Hmm, I am confused, how would that be better?

Comment (jblefkove-va) [reply us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:103]: it eliminates the first check

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+                e ->
+                    StringUtils.isNotEmpty(e.value())
+                        && StringUtils.isNotEmpty(e.system())
+                        && !"http://hl7.org/fhir/sid/us-npi".equals(e.system()));
+
+    boolean validNpiSlice =
+        identifier.stream().filter(e -> "http://hl7.org/fhir/sid/us-npi".equals(e.system())).count()
+            == 1;
+
+    return (validAllSlice ^ validNpiSlice);
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:112]: It's less clear to me now what this method is checking? Are these the conditions?

1. At least one identifier is present
2. Every identifier has system and value set, unless system is NPI
3. At most one identifier has system NPI

If so, I think it is simpler to keep the conditions separate and return early if necessary, i.e:

1. Return false if there are no identifiers (method already does this)
2. Return false if there is more than one NPI identifier
3. Return true/false if every identifier has system and value set, filtering out those where system is NPI

Comment (bschofield-va) [reply us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:112]: 10/10 agree, this fails Occam's razor.
I value simple to read code. No points for cleverness.

File: us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+        > 1) {
+      return false;
+    }
+
+    return identifier.stream()
+        .allMatch(
+            e ->
+                StringUtils.isNotEmpty(e.system())
+                    && (StringUtils.isNotEmpty(e.value())
+                        || "http://hl7.org/fhir/sid/us-npi".equals(e.system())));
```
Comment (jblefkove-va) [inline us-core-r4/src/main/java/gov/va/api/health/r4/api/resources/Practitioner.java:110]: You can simplify this check by filtering out the NPI entries first:
```
    return identifier
        .stream()
        .filter(e -> !"http://hl7.org/fhir/sid/us-npi".equals(e.system()))
        .allMatch(e -> isNotEmpty(e.system()) && isNotEmpty(e.value()));
```

(Also, we are not fully consistent about it, but we often statically import from `StringUtils`. In this case that definitely benefits readability.)

**Review by jblefkove-va — APPROVED** [review]

**Review by ilaflamme-va — APPROVED** [review]

**Pull request merged**